File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/giftcard/GiftCard.java
package it.giannibombelli.wsc2026.giftcard.domain.giftcard;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.model.Aggregate;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.events.*;

import java.util.Objects;

public final class GiftCard implements Aggregate<GiftCardId> {
    private final GiftCardId id;
    private Money balance;

    public GiftCard(GiftCardId id, Money balance) {
        Require.requireArgument(id, "id");
        Require.requireArgument(balance, "balance");
        this.id = id;
        this.balance = balance;
    }

    public GiftCardId id() {
        return id;
    }

    public Money balance() {
        return balance;
    }

    public static GiftCard issue(GiftCardId id) {
        Require.requireArgument(id, "id");
        return new GiftCard(id, Money.zero());
    }

    public GiftCardTopUpRequested requestTopUp(Money amount) {
        Require.requireArgument(amount, "amount");
        return new GiftCardTopUpRequested(this.id, amount, this.balance);
    }

    public GiftCardRedeemEvents redeem(Money amount) {
        Require.requireArgument(amount, "amount");
        if (this.balance.isLessThan(amount)) {
            return new GiftCardRedeemEvents.GiftCardNotRedeemed(this.id, amount,
                new Description("insufficient balance"));
        }
        this.balance = this.balance.minus(amount);
        return new GiftCardRedeemEvents.GiftCardRedeemed(this.id, amount);
    }

    public GiftCardRefunded refund(Money amount) {
        Require.requireArgument(amount, "amount");
        this.balance = this.balance.plus(amount);
        return new GiftCardRefunded(this.id, amount);
    }

    public TopUpConfirmed confirmTopUp(Money amount) {
        Require.requireArgument(amount, "amount");
        this.balance = this.balance.plus(amount);
        return new TopUpConfirmed(this.id, amount);
    }

    public GiftCardCredited credit(Money amount) {
        Require.requireArgument(amount, "amount");
        this.balance = this.balance.plus(amount);
        return new GiftCardCredited(this.id, amount);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return Objects.equals(id, ((GiftCard) o).id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "GiftCard[" +
            "id=" + id + ", " +
            "balance=" + balance +
            ']';
    }
}

File: typescript/src/giftcard/domain/giftcard/giftCard.ts
// Aggregato GiftCard: emette e gestisce il saldo di una gift card.

import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { GiftCardCredited, giftCardCredited } from '../events/giftCardCredited.js';
import type { GiftCardRedeemEvent } from '../events/giftCardRedeemEvents.js';
import { giftCardRedeemed, giftCardNotRedeemed } from '../events/giftCardRedeemEvents.js';
import { GiftCardRefunded, giftCardRefunded } from '../events/giftCardRefunded.js';
import { GiftCardTopUpRequested, giftCardTopUpRequested } from '../events/giftCardTopUpRequested.js';
import { TopUpConfirmed, topUpConfirmed } from '../events/topUpConfirmed.js';
import { GiftCardId } from './giftCardId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class GiftCard {
  private readonly _id: GiftCardId;
  private _balance: Money;

  constructor(id: GiftCardId, balance: Money) {
    requireArgument(id, 'id');
    requireArgument(balance, 'balance');
    this._id = id;
    this._balance = balance;
  }

  id(): GiftCardId {
    return this._id;
  }

  balance(): Money {
    return this._balance;
  }

  static issue(id: GiftCardId): GiftCard {
    requireArgument(id, 'id');
    return new GiftCard(id, Money.zero());
  }

  requestTopUp(amount: Money): GiftCardTopUpRequested {
    requireArgument(amount, 'amount');
    return giftCardTopUpRequested(this._id, amount, this._balance);
  }

  redeem(amount: Money): GiftCardRedeemEvent {
    requireArgument(amount, 'amount');
    if (this._balance.isLessThan(amount)) {
      return giftCardNotRedeemed(this._id, amount, new Description('insufficient balance'));
    }
    this._balance = this._balance.minus(amount);
    return giftCardRedeemed(this._id, amount);
  }

  refund(amount: Money): GiftCardRefunded {
    requireArgument(amount, 'amount');
    this._balance = this._balance.plus(amount);
    return giftCardRefunded(this._id, amount);
  }

  confirmTopUp(amount: Money): TopUpConfirmed {
    requireArgument(amount, 'amount');
    this._balance = this._balance.plus(amount);
    return topUpConfirmed(this._id, amount);
  }

  credit(amount: Money): GiftCardCredited {
    requireArgument(amount, 'amount');
    this._balance = this._balance.plus(amount);
    return giftCardCredited(this._id, amount);
  }
}
